### PR TITLE
Add support for Crystalfontz CFAF7201280A0_050Tx modules

### PR DESCRIPTION
**Pull request by MWP**
Adding support for upcoming new product lines from Crystalfontz America Inc.
These compact modules host a Raspberry Pi CM4 and feature a 7in 720x1280 TFT panel (with or without cap-touch), extended GPIO, battery backed-up clock, and other features.
These products have completed beta testing and will go into production early/mid 2023.

### CODE REVIEW

File: arch/arm/boot/dts/overlays/README [hunk 1]
```diff
@@ -801,6 +801,14 @@ Params: cma-512                 CMA is 512MB (needs 1GB)
         cma-default             Use upstream's default value
 
 
+Name:   crystalfontz-cfaf7201280a0_050tx
+Info:   Overlay to configure the Crystalfontz CFA7201280A0_050TN and
+        CFA7201280A0_050TC 7" non-touchscreen and touchscreen (respectively)
+		Raspberry Pi CM4 based modules.
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:807]: No TAB characters please - only spaces are allowed for indent.

File: arch/arm/boot/dts/overlays/README [hunk 2]
```diff
@@ -801,6 +801,14 @@ Params: cma-512                 CMA is 512MB (needs 1GB)
         cma-default             Use upstream's default value
 
 
+Name:   crystalfontz-cfaf7201280a0_050tx
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:804]: Catchy name.

File: arch/arm/boot/dts/overlays/crystalfontz-cfaf7201280a0_050tx.dts [hunk 3]
```diff
@@ -0,0 +1,119 @@
+/*
+ * crystalfontz-cfaf7201280a0_050tx.dts
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/crystalfontz-cfaf7201280a0_050tx.dts:2]: Please include `-overlay` in the filename, otherwise the wrong Makefile rule is used.

**Review by pelwell — COMMENTED** [review]
This looks OK aside from the inline comments.

File: arch/arm/boot/dts/overlays/README (hunk 2 above)
Comment (6by9) [reply arch/arm/boot/dts/overlays/README:804]: I had been trying to group all the DSI displays under vc4-kms-dsi-* to emphasize that they will only work alongside vc4-kms-v3d. (Likewise DPI was under vc4-kms-dpi-*).

We did miss that with cutiepi-panel, but IMHO would be preferable.

File: arch/arm/boot/dts/overlays/crystalfontz-cfaf7201280a0_050tx.dts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		};
+	};
+// GT928 TOUCH CONTROLLER IC
+	fragment@4 {
+		target = <&i2c_csi_dsi>;
+		__overlay__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+			gt928@5d {
```
Comment (6by9) [inline arch/arm/boot/dts/overlays/crystalfontz-cfaf7201280a0_050tx.dts:94]: Your description says that this overlay supports both a touchscreen and non-touchscreen variant, but the touch driver is always enabled.
The Goodix driver does check an ID register, but it'd be nice to have an override to explicitly disable it.

File: drivers/gpu/drm/panel/panel-ilitek-ili9881c.c [hunk 5]
```diff
@@ -842,6 +1050,20 @@ static const struct drm_display_mode nwe080_default_mode = {
...
+	.hsync_end	= 720 + 87 + 20, 
+	.htotal		= 720 + 87 + 20 + 87,
+	.vdisplay	= 1280,
+	.vsync_start= 1280 + 16,
+	.vsync_end	= 1280 + 16 + 8,
+	.vtotal		= 1280 + 16 + 8 + 16,
+	.width_mm	= 62,
+	.height_mm	= 1108
+};
+
```
Comment (6by9) [inline drivers/gpu/drm/panel/panel-ilitek-ili9881c.c:1066]: Indentation here has gone wrong. Kernel coding style is tabs at 8 spaces per tab.

File: drivers/gpu/drm/panel/panel-ilitek-ili9881c.c [hunk 6]
```diff
@@ -675,6 +676,213 @@ static const struct ili9881c_instr nwe080_init[] = {
...
+	ILI9881C_COMMAND_INSTR(0xCD, 0x18),
+	ILI9881C_COMMAND_INSTR(0xCE, 0x4C),
+	ILI9881C_COMMAND_INSTR(0xCF, 0x33),
+	ILI9881C_COMMAND_INSTR(0xD0, 0x27),
+	ILI9881C_COMMAND_INSTR(0xD1, 0x4B),
+	ILI9881C_COMMAND_INSTR(0xD2, 0x60),
+	ILI9881C_COMMAND_INSTR(0xD3, 0x39),
+};
+
+
```
Comment (6by9) [inline drivers/gpu/drm/panel/panel-ilitek-ili9881c.c:885]: Kernel coding style says not to use double blank lines.

File: drivers/gpu/drm/panel/panel-ilitek-ili9881c.c [hunk 7]
```diff
@@ -728,10 +936,10 @@ static int ili9881c_prepare(struct drm_panel *panel)
 	msleep(5);
 
 	/* And reset it */
-	gpiod_set_value(ctx->reset, 1);
+	gpiod_set_value_cansleep(ctx->reset, 1);
```
Comment (6by9) [inline drivers/gpu/drm/panel/panel-ilitek-ili9881c.c:939]: Switching to gpiod_set_value_cansleep is not directly part of adding support for your panel, and therefore should be a separate patch.

**Review by 6by9 — COMMENTED** [review]
Thanks for the PR.

Is this for a product with integrated CM4, or a standalone product? You've only added the backlight driver to bcm2711_defconfig, but a generic product could run on CM1 or CM3.

Are Crystalfontz the panel manufacturer or an integrator? The panel compatible should use the panel manufacturer. (Overlay name is reasonable to be the integrator as it includes the GPIO assignments that they define).

No Signed-off-by: tag on the patches.

This should be 4 patches:
- Switch panel driver to gpiod_set_value_cansleep
- Add compatible to panel driver
- defconfig
- overlay

Ideally there should be a 5th which adds the compatible to the device tree bindings too.

**Comment by pelwell** [comment]
CM4 is a totally different form factor to earlier Compute Modules, so a product that hosts a CM4 (i.e. a CM4 carrier) won't run on anything but a CM4, making the restriction to BCM2711 defconfigs reasonable (if not required).

**Comment by 6by9** [comment]
> CM4 is a totally different form factor to earlier Compute Modules, so a product that hosts a CM4 (i.e. a CM4 carrier) won't run on anything but a CM4, making the restriction to BCM2711 defconfigs reasonable (if not required).

I'd missed that initial comment that it was a module to host a CM4, hence my question over generic or not.
If it had a 22pin 0.05mm FFC connector (which it may still have), then it would equally work on a CM1/3.

**Comment by failure132** [comment]
Dear Linux,

How should I say this without hurting you. It's not you it's me, I'm breaking up with you. Please leave! Yay! Grow up and send a message stop being a child, what am I going to do call the police? Go hire a lawyer no obviously idiot. I'm over there driving to the ends of the earth for some dead girl.... Supposedly and your over there clacking away like a clacker. Ugh Tracy Flores, or lorde or whoever the hell you are. Get it together stop wasting my time

**Comment by MWP** [comment]
Thanks all, this is my first kernel pull request (as you may have guessed).

> Is this for a product with integrated CM4, or a standalone product? You've only added the backlight driver to bcm2711_defconfig, but a generic product could run on CM1 or CM3.

It has an integrated CM4.
These Crystalfontz modules will not support CM1-CM3.

> Are Crystalfontz the panel manufacturer or an integrator? The panel compatible should use the panel manufacturer. (Overlay name is reasonable to be the integrator as it includes the GPIO assignments that they define).

Crystalfontz are both a panel manufacturer and an integrator.
The panel is also sold separately by Crystalfontz and I don't believe it is available elsewhere.
 
> No Signed-off-by: tag on the patches.

Ok.

> This should be 4 patches:
> * Switch panel driver to gpiod_set_value_cansleep
> * Add compatible to panel driver
> * defconfig
> * overlay

Ok, but can you please clear something up for me?
Does that mean 4 commits on a single pull req, or 4 separate pull req's?
Should I delete this pull req, and create a new one with the corrections?

> Ideally there should be a 5th which adds the compatible to the device tree bindings too.

Ok.

**Comment by popcornmix** [comment]
> Does that mean 4 commits on a single pull req, or 4 separate pull req's?

4 commits in one PR

> Should I delete this pull req, and create a new one with the corrections?

No. Just rejig your commits on your current branch, and when they look as desired, "git push -f" will update this PR.
